File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/viewmodel/ProfileDetailsViewModel.kt
package com.perrystreet.woof.presentation.profile.viewmodel

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.presentation.common.viewmodel.StateProducingViewModel
import com.perrystreet.woof.presentation.profile.mapper.DogProfileDomainToUIModelMapper
import com.perrystreet.woof.presentation.profile.uimodel.ProfileContentUIModel
import com.perrystreet.woof.presentation.profile.uimodel.ProfileDetailsUIModel
import com.perrystreet.woof.usecase.dogs.GetDogProfileUseCase
import io.reactivex.rxjava3.kotlin.plusAssign
import org.koin.android.annotation.KoinViewModel
import org.koin.core.annotation.InjectedParam
import java.util.Optional

@KoinViewModel
class ProfileDetailsViewModel(
    @InjectedParam private val dog: Dog,
    private val getDogProfileUseCase: GetDogProfileUseCase,
    private val profileMapper: DogProfileDomainToUIModelMapper,
) : StateProducingViewModel<ProfileDetailsUIModel>(
    initialValue = ProfileDetailsUIModel(
        name = dog.name,
        summary = null,
        heroTags = emptyList(),
        content = ProfileContentUIModel.Loading,
    ),
) {
    override fun onFirstAppear() {
        disposables += getDogProfileUseCase(dog).subscribe(
            { profile -> _state.onNext(profileMapper(profile)) },
            { error -> mutableError.onNext(Optional.of(error)) },
        )
    }
}

File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/ViewModel/ProfileDetailsViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase
import Utils

@Factory
public final class ProfileDetailsViewModel: StateProducingViewModel<ProfileDetailsUIModel, DogsError> {
    @DIArgument private let dog: Dog
    private let getDogProfileUseCase: GetDogProfileUseCase
    private let profileMapper: DogProfileDomainToUIModelMapper

    init(
        dog: Dog,
        getDogProfileUseCase: GetDogProfileUseCase,
        profileMapper: DogProfileDomainToUIModelMapper
    ) {
        self.dog = dog
        self.getDogProfileUseCase = getDogProfileUseCase
        self.profileMapper = profileMapper
        super.init(
            initialValue: ProfileDetailsUIModel(
                name: dog.name,
                summary: nil,
                heroTags: [],
                content: .loading
            )
        )
    }

    public override func onFirstAppear() {
        getDogProfileUseCase(dog: dog)
            .pss_sink { [weak self] result in
                guard let self else { return }
                switch result {
                case .success(let profile): setState(profileMapper(profile: profile))
                case .failure(let error): emitError(error)
                }
            }
            .store(in: &cancellables)
    }
}
